### PR TITLE
fix: harden MCP refresh against duplicate tool failures

### PR DESCRIPTION
**Pull request by sathiya06**
Deduplicate duplicate discovered tool names and rollback failed transactions before persisting refresh errors so integrations never remain stuck in REFRESHING with a null error after sync failures.

## Description

When an MCP server advertises the same tool name more than once in `tools/list`, integration refresh attempted to insert duplicate tool rows sharing the same `namespaced_name`, causing a database constraint violation and HTTP 409/500 responses. Worse, the error handler reused a session whose transaction had already been aborted (`PendingRollbackError`), so the failure was never persisted — the integration stayed in `refresh_status: refreshing` with a null `refresh_error` and no way for the UI to show what went wrong.

This change deduplicates duplicate discovered tool names during sync (first occurrence wins; later duplicates are skipped with a warning) and rolls back the session in `_persist_refresh_error` before writing the ERROR state. Per the [MCP tool names specification](https://modelcontextprotocol.io/specification/2026-07-28/server/tools#tool-names), tool names **SHOULD** be unique within a server; this makes Syntara resilient when a server violates that guidance.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactoring (no functional changes)
- [ ] Performance improvement
- [x] Test coverage improvement

## Changes Made

- [x] Skip duplicate tool names during MCP tool sync (first occurrence wins; log a warning)
- [x] Roll back the DB session in `_persist_refresh_error` before persisting ERROR state
- [x] Add integration tests for duplicate-name deduplication and error-state persistence after sync failures

## Out of Scope

- Rejecting refresh outright when duplicate names are detected (dedupe is preferred so partially valid tool lists still sync)
- Changing global IntegrityError handling or HTTP status mapping
- Frontend changes to display refresh errors (backend fix only)

## Related Issues

- AAP-87199

## How to Test

1. Point an MCP integration at a server whose `tools/list` returns the same tool name twice.
2. `POST /integrations/{id}/refresh` — refresh should succeed; only distinct tools are synced.
3. To verify error persistence: force a sync-time DB constraint failure (e.g. colliding `namespaced_name`) and confirm the integration ends in `refresh_status: error` with a non-null `refresh_error` instead of an endless refreshing spinner.

Automated:

```bash
cd backend
uv run pytest tests/integration/integrations/test_integration_refresh.py -v
```

## Backend Checklist

- [x] I have run `make test` and all tests pass (in `backend/`) — `tests/integration/integrations/test_integration_refresh.py` (11/11)
- [ ] I have run `make lint` and code passes all checks
- [ ] I have run `make format` to ensure consistent formatting
- [ ] I have run `make typecheck` and all types are correct
- [x] I have added tests for new functionality
- [ ] Database migrations are included if schema changed
- [x] No sensitive information (keys, passwords, etc.) is included

## Frontend Checklist

- [ ] I have run `npm test` and all tests pass (in `frontend/`)
- [ ] I have run `npm run lint` and code passes all checks
- [ ] I have run `npm run tsc` and there are no type errors
- [ ] I have added tests (unit / E2E) for new functionality
- [ ] Accessibility has been considered (semantic HTML, labels, keyboard navigation)
- [ ] Screenshots or screen recordings are attached for UI changes

## Visual Regression

> Visual regression does **not** run automatically on this PR. If your change affects UI that's covered by `e2e/visual-regression/page-registry.ts`, update baselines yourself before requesting review — otherwise the weekly baseline-refresh PR will pick up the drift and route it to the UI/UX team instead of you.

- [ ] If this PR intentionally changes covered UI: I have commented `/update-screenshots` on this PR and confirmed the regenerated baselines look correct in the Files changed tab
- [ ] If this PR adds a new page/route: I have added an entry to `page-registry.ts` and run `/update-screenshots` to generate its baseline

## General Checklist

- [x] Code follows the project's coding conventions
- [ ] I have updated relevant documentation
- [x] No secrets or credentials are included in this PR
- [ ] Breaking changes are documented with migration steps

## Screenshots / Demo (if applicable)

_N/A — backend-only change._

## Additional Notes

The MCP spec states that tool names **SHOULD** be unique within a server ([Tool Names](https://modelcontextprotocol.io/specification/2026-07-28/server/tools#tool-names)). Clients aggregating multiple servers may still see cross-server collisions and need disambiguation, but duplicate names within a single server's `tools/list` response are malformed; Syntara now handles that gracefully instead of wedging the integration.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by Rishi625 — APPROVED** [review]
lgtm... few minor sugggestions bellow

File: backend/tests/integration/integrations/test_integration_refresh.py [hunk 1]
```diff
@@ -221,6 +222,97 @@ async def test_refresh_marks_missing_tools_without_disabling(
         assert tools["alpha"].enabled is True
         assert tools["alpha"].status == ToolStatus.AVAILABLE
 
+    async def test_refresh_dedupes_duplicate_tool_names(
+        self, auth_client: AsyncClient, test_db_session: AsyncSession, test_user: User
+    ) -> None:
+        """A server advertising the same tool name twice is deduped, not 500'd.
+
+        Regression for AAP-87199: two DiscoveredTools with the same name used to
```
Comment (Rishi625) [inline backend/tests/integration/integrations/test_integration_refresh.py]: Nit: this cites AAP-87199, but the PR body Related Issues section says none. Worth linking the ticket there. Also folllow the upstream rules, do not mention internal references in the committed files

File: backend/tests/integration/integrations/test_integration_refresh.py [hunk 2]
```diff
@@ -221,6 +222,97 @@ async def test_refresh_marks_missing_tools_without_disabling(
...
+            response = await auth_client.post(f"{BASE_URL}/{integration_id}/refresh")
+
+        assert response.status_code == 200
+        data = response.json()
+        # Only two distinct tools synced despite three discovered entries.
+        assert data["synced_count"] == 2
+
+        tools = (await test_db_session.exec(select(Tool).where(Tool.integration_id == UUID(integration_id)))).all()
+        assert {t.name for t in tools} == {"dup", "unique"}
+        assert len([t for t in tools if t.name == "dup"]) == 1
```
Comment (Rishi625) [inline backend/tests/integration/integrations/test_integration_refresh.py:257]: Nit: first-wins is the stated contract (first `dup` has params, second does not), but this only asserts row count/name. A parameter check on the kept `dup` row would lock that in.

File: backend/src/syntara/integrations/services/integration_service.py [hunk 3]
```diff
@@ -790,6 +790,12 @@ async def _persist_refresh_error(self, integration_id: UUID, exc: Exception) ->
             integration_id=str(integration_id),
             error_type=type(exc).__name__,
         )
+        # The triggering exception may have left the session in a failed
+        # transaction (e.g. an IntegrityError raised by flush). Roll back before
+        # issuing new statements, otherwise the SELECT in _get_or_raise raises
+        # PendingRollbackError and the ERROR state is never persisted, leaving
+        # the integration wedged in refresh_status=REFRESHING with a null error.
+        await self.session.rollback()
```
Comment (Rishi625) [inline backend/src/syntara/integrations/services/integration_service.py:798]: Nit / follow-up: this rollback is the right fix for refresh. `validate_integration`'s unexpected-error path still does `_get_or_raise` without rolling back first — same pattern if a flush abort ever happens there.

File: backend/tests/integration/integrations/test_integration_refresh.py (hunk 1 above)
Comment (sathiya06) [reply backend/tests/integration/integrations/test_integration_refresh.py]: Will remove it from the code

Comment (sathiya06) [reply backend/tests/integration/integrations/test_integration_refresh.py]: Removed AAP-87199 from the test docstrings; the ticket is linked in the PR Related Issues section instead.

File: backend/tests/integration/integrations/test_integration_refresh.py (hunk 2 above)
Comment (sathiya06) [reply backend/tests/integration/integrations/test_integration_refresh.py:257]: Added ToolParameter assertions on the kept dup row to lock in first-wins behavior (query param from the first occurrence).

File: backend/src/syntara/integrations/services/integration_service.py (hunk 3 above)
Comment (sathiya06) [reply backend/src/syntara/integrations/services/integration_service.py:798]: Added session.rollback() before _get_or_raise in validate_integration's unexpected-error path, matching _persist_refresh_error.

**Review by BrennanPaciorek — APPROVED** [review]

**Review by BrennanPaciorek — COMMENTED** [review]
One finding is some missing coverage:
```
  Missing coverage

  validate_integration's unexpected-error rollback is untested.

  - Evidence:
  except Exception as exc:
      ...
  +     await self.session.rollback()
        integration = await self._get_or_raise(integration_id, for_update=True)
        integration.validation_status = IntegrationStatus.ERROR
  - This is the identical fix pattern applied to _persist_refresh_error, added specifically in response to reviewer Rishi625's comment: "validate_integration's unexpected-error path still does _get_or_raise without rolling
  back first — same pattern if a flush abort ever happens there." The author's reply confirms the fix was added — but no test was added alongside it.
  - I checked the existing test suite for anything that would exercise this branch: test_validate_integration_failure_emits_error_event and test_validate_integration_failure_preserves_error_message (in
  test_integration_audit_events.py) both mock MCPServerAdapter.validate() to return ValidateResult(success=False, ...) — a handled result, not a raised exception. Neither ever enters the except Exception block where the new
  rollback line lives, so the rollback is currently exercised by zero tests.
  - Confidence: HIGH — confirmed by reading the actual except-block code on the PR branch and the two closest existing tests' mock setup.
  - Why it matters: this is the exact bug class the PR is fixing (silent wedge in a status field after a transaction abort) — for _persist_refresh_error it got a dedicated regression test that manufactures a real DB
  constraint violation; the sibling fix in validate_integration did not. If that line were ever wrong (wrong exception scope, wrong session, wrong ordering), nothing in CI would catch it, and validation would silently regress
  to the same wedging behavior this PR exists to fix.
  - Suggested test: mirror test_refresh_persists_error_when_sync_raises for validate_integration — pre-seed a colliding row or patch a step inside the try block to raise after a flush, call validate_integration, and assert
  validation_status == ERROR with a non-null validation_error (rather than the call raising PendingRollbackError before recording anything).
```

**Comment by sathiya06** [comment]
Addressed the missing-coverage finding: added `test_validate_persists_error_when_commit_aborts` in `test_integration_validate.py`. It mirrors `test_refresh_persists_error_when_sync_raises` — mock validate success, abort the transaction on the 2nd `session.commit` with `IntegrityError`, and assert the integration ends in `validation_status=ERROR` with a non-null `validation_error` (409 response).
